test/facility-covenant/patch-facility-covenants.api-test.ts

```diff
@@ -0,0 +1,364 @@
+
+      const { status, body } = await api.patch(updateFacilityCovenantUrl, requestBodyToUpdateFacilityCovenant);
+
+      expect(status).toBe(400);
+      expect(body).toStrictEqual({ message: 'Bad request', error: acbsErrorMessage, statusCode: 400 });
+    });
+
+    it('returns a 400 response if ACBS responds with a 400 response that is a string that does not contain "Facility not found" when replacing the facility covenant', async () => {
+      const acbsErrorMessage = 'ACBS error message';
+      requestToReplaceFacilityCovenantWithBody(updatedCovenants[1]).reply(400, acbsErrorMessage);
+
+      const { status, body } = await api.patch(updateFacilityCovenantUrl, requestBodyToUpdateFacilityCovenant);
+
+      expect(status).toBe(400);
+      expect(body).toStrictEqual({ message: 'Bad request', error: acbsErrorMessage, statusCode: 400 });
+    });
+
+    it('returns a 500 response if ACBS responds with an error code that is not 400 when replacing the facility covenant', async () => {
+      requestToReplaceFacilityCovenantWithBody(updatedCovenants[1]).reply(401, 'Unauthorized');
+
+      const { status, body } = await api.patch(updateFacilityCovenantUrl, requestBodyToUpdateFacilityCovenant);
+
+      expect(status).toBe(500);
+      expect(body).toStrictEqual({ message: 'Internal server error', statusCode: 500 });
+    });
+
+    it('returns a 500 response if ACBS times out when creating the facility covenant', async () => {
+      requestToReplaceFacilityCovenantWithBody(updatedCovenants[1]).delay(TIME_EXCEEDING_ACBS_TIMEOUT).reply(201);
+
+      const { status, body } = await api.patch(updateFacilityCovenantUrl, requestBodyToUpdateFacilityCovenant);
+
+      expect(status).toBe(500);
+      expect(body).toStrictEqual({
+        statusCode: 500,
+        message: 'Internal server error',
+      });
+    });
+  });
+
+  describe('field validation', () => {
+    const makeRequest = (body: unknown[]) => api.patch(updateFacilityCovenantUrl, body);
+    const givenAnyRequestBodyWouldSucceed = () => {
+      givenAuthenticationWithTheIdpSucceeds();
+      givenRequestToGetFacilityCovenantsFromAcbsSucceeds();
+      givenAnyRequestBodyToReplaceFacilityCovenantInAcbsSucceeds();
+      givenAnyRequestBodyToReplaceFacilityCovenantInAcbsSucceeds();
+      givenAnyRequestBodyToReplaceFacilityCovenantInAcbsSucceeds();
+    };
+
+    withNonNegativeNumberFieldValidationApiTests({
+      fieldName: 'targetAmount',
+      required: false,
+      validRequestBody: requestBodyToUpdateFacilityCovenant,
+      makeRequest,
+      givenAnyRequestBodyWouldSucceed,
+    });
+
+    withDateOnlyFieldValidationApiTests({
+      fieldName: 'expirationDate',
+      required: false,
+      nullable: false,
+      validRequestBody: requestBodyToUpdateFacilityCovenant,
+      makeRequest,
+      givenAnyRequestBodyWouldSucceed,
+    });
+
+    it('returns a 400 response if the request body is the empty object', async () => {
```

Comment:
Could we also test:
- if the request body is an array
- if the request body is a string
- if the request body is a number
- if the request body is the literal value null
- if the request body is true or false
please?

For everything other than an object / array, it's okay that the response message will be different, I just want to be confident that they always return a 400 error code (i.e. you can just test for the 400 status code and ignore the rest of the message apart from for array and object)

Reply:
Ah interesting. I wonder if it's expecting us to `JSON.stringify` beforehand if we want to send non-object values?

E.g. `JSON.stringify(2)`, `JSON.stringify(true)`, `JSON.stringify("a string")`, etc.

Reply:
This works, thank you!